### PR TITLE
MySQL driver+test => utf8mb3 -> utf8mb4

### PR DESCRIPTION
**Pull request by inoas**
- [x] Enhancement
- [x] Bug (inserting valid utf8 such as 😃 will break your data using mysql/cakephp and default utf8 settings set by the driver and recommended app.php)

Refs:
* issue with more info: https://github.com/cakephp/cakephp/issues/11324#issuecomment-336672560
* related app skeleton PR https://github.com/cakephp/app/pull/554
* super-cedes https://github.com/cakephp/cakephp/pull/11328 (fixes unit tests)

### CODE REVIEW

**Comment by inoas** [comment]
Is it a good idea to add a unit test that tests this though a MySQL connection?:

```sql
DROP TABLE IF EXISTS `utf8mb4test`;
CREATE TABLE `utf8mb4test` (
  `field` TEXT CHARACTER SET utf8mb4 NOT NULL
) ENGINE=INNODB;

SET NAMES utf8;
INSERT INTO utf8mb4test SET field = '😃'; # [ERROR IN QUERY 2] Incorrect STRING VALUE: '\xF0\x9F\x98\x83' FOR COLUMN 'field' AT ROW 1

SET NAMES utf8mb4;
INSERT INTO utf8mb4test SET field = '😃'; # Should work
```

If yes, where would it go and how should I do it? Is pure SQL against the driver, including the create table statement, ok?

**Comment by inoas** [comment]
I'd need a little bit of assistance with the unit tests. I don't know how to select the test database...

**Comment by chinpei215** [comment]
The `cake` database seems to be used for Mock Connection. So I think you cannot connect the database actually. You may have to choose one database in the following list:
https://github.com/cakephp/cakephp/blob/39d7fa96eaeaafb89f214f462878b669a832f9fa/.travis.yml#L49-L51

**Comment by inoas** [comment]
Hm... locally I get a PDOException.

**Comment by chinpei215** [comment]
@inoas What sql_mode are you using on your database server?

**Comment by inoas** [comment]
Do you mean this?
```sql
# In
SELECT @@SQL_MODE
# Out
STRICT_TRANS_TABLES,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION
```

**Comment by inoas** [comment]
Let me know if this is any good, then I will squash all commits.

File: tests/TestCase/Database/Driver/MysqlTest.php [hunk 1]
```diff
@@ -163,4 +163,138 @@ public function testCommitTransactionAutoConnect()
...
+
+        $connection = new \Cake\Database\Connection([
+            'driver' => '\Cake\Database\Driver\Mysql',
+            'persistent' => true,
+            'host' => 'localhost',
+            'username' => 'root',
+            'password' => '',
+            'database' => 'cakephp_test',
+            'port' => '3306',
+            // Try to insert utf8 4-byte symbol through utf8 3-byte connection character set
```
Comment (markstory) [inline tests/TestCase/Database/Driver/MysqlTest.php]: What is this test testing? It looks like it is testing MySQL, I'm not sure we need to do that. The tests above ensure that the DSN sent to PDO is correctly built which is enough for me.

Comment (inoas) [reply tests/TestCase/Database/Driver/MysqlTest.php]: Then I assume correctly that none of the new tests are required for this PR to become mergable, correct?

Comment (markstory) [reply tests/TestCase/Database/Driver/MysqlTest.php]: I don't think this test is necessary at least.

Comment (inoas) [reply tests/TestCase/Database/Driver/MysqlTest.php]: Well the DSN is tested above by another unit. I will remove all the other tests.

**Pull request merged**